File: JS/Tree/144.Binary Tree Preorder Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {  // Recursive preOrder
    const result = [];
    (function preorder(node) {
        if (node) {
            result.push(node.val);
            preorder(node.left);
            preorder(node.right);
        }
    })(root);
    return result;
};

var preorderTraversal = function(root) { // Stack preOrder
    const result = [];
    const stack = [root];
    while(stack.length) {
        const node = stack.pop();
        if (node) {
            result.push(node.val);
            stack.push(node.right);
            stack.push(node.left);
        }
    }
    return result;
};

var preorderTraversal = function(root) { // Morris traversal
    const result = [];
    let current = root;
    let previous;
    while(current) {
        if (!current.left) {
            result.push(current.val);
            current = current.right;
        } else {
            previous = current.left;
            while(previous.right && previous.right !== current) {
                previous = previous.right;
            }
            if (!previous.right) {
                previous.right = current;
                result.push(current.val);
                current = current.left;
            } else {
                previous.right = null;
                current = current.right;
            }
        }
    }
    return result;
};

File: Golang/Tree/144.binary-tree-preorder-traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func postorderTraversal(root *TreeNode) []int {
	ret := []int{}
	var recursive func(node *TreeNode)
	recursive = func(node *TreeNode) {
		if node != nil {
			recursive(node.Left)
			recursive(node.Right)
			ret = append(ret, node.Val)
		}
	}
	recursive(root)
	return ret
}

func postorderTraversalPointer(root *TreeNode) []int {
	ret := []int{}
	if root != nil {
		stack := []*TreeNode{root, root}
		for len(stack) > 0 {
			node := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if len(stack) > 0 && node == stack[len(stack)-1] {
				if node.Right != nil {
					stack = append(stack, node.Right)
					stack = append(stack, node.Right)
				}
				if node.Left != nil {
					stack = append(stack, node.Left)
					stack = append(stack, node.Left)
				}
			} else {
				ret = append(ret, node.Val)
			}
		}
	}
	return ret
}
